www/caddy/src/opnsense/www/js/widgets/CaddyCertificate.js

```diff
@@ -0,0 +1,143 @@
+ * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import BaseTableWidget from "./BaseTableWidget.js";
+
+export default class CaddyCertificate extends BaseTableWidget {
+    constructor() {
+        super();
+        this.resizeHandles = "e, w";
+        this.currentCertificates = {};
+
+        // Since we only update when dataHasChanged we can almost update in real time
+        this.tickTimeout = 2000;
+
+    }
+
+    getGridOptions() {
+        return {
+            // trigger overflow-y:scroll after 650px height
+            sizeToContent: 650
+        };
+    }
+
+    getMarkup() {
+        let $container = $('<div></div>');
+        let $caddyCertificateTable = this.createTable('caddyCertificateTable', {
+            headerPosition: 'none'
+        });
+
+        $container.append($caddyCertificateTable);
+        return $container;
+    }
+
+    async onWidgetTick() {
+        try {
+            // Check if Caddy is enabled
+            const caddyStatus = await ajaxGet('/api/caddy/reverse_proxy/get', {});
+            if (!caddyStatus.caddy.general || caddyStatus.caddy.general.enabled === "0") {
+                this.displayError(`${this.translations.unconfigured}`);
+                return;
+            }
+
+            // Fetch the certificate details
+            const response = await ajaxGet('/api/caddy/diagnostics/certificate', {});
+            if (response.status !== "success") {
+                this.displayError(`${this.translations.nocerts}`);
+                return;
+            }
+
+            // Process certificates if the response is successful
+            this.processCertificates(response.content);
+        } catch (error) {
+            this.displayError(`${this.translations.error}`);
+        }
+    }
+
+    // Utility function to display errors within the widget
+    displayError(message) {
+        const $error = $(`<div class="error-message"><a href="/ui/caddy/general">${message}</a></div>`);
+        $('#caddyCertificateTable').empty().append($error);
+    }
+
+    // Checks if the data has changed to prevent unnecessary UI updates
+    dataHasChanged(newCertificates) {
+
+        // Directly serialize the entire newCertificates array to perform a deep comparison
+        const newCertificatesString = JSON.stringify(newCertificates);
+        const currentCertificatesString = JSON.stringify(this.currentCertificates);
+
+        // Compare the serialized strings
+        if (newCertificatesString !== currentCertificatesString) {
```

Comment:
Just an idea... perhaps hashes (md5) of the strings might be quicker than serializing them to json first and then doing a comparison operation, the comparison between md5 hashes is way faster also. Probably easy to benchmark as well.

The operation applies to all widgets that poll for new data, i think the basewidget should provide an easy to use function for it.. let me think about it

Reply:
Comparing hashes sounds like a great idea for a generalized base function.

Reply:
I would like to leave it the way it is right now.